### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by klmedeiros-ag**
Functional work with better assert that messages, setAttribute fixed and sorted, and proper printing.

### CODE REVIEW

File: R/annotate.R [hunk 1]
```diff
@@ -1,67 +1,78 @@
...
+#' @importFrom stats setNames
 #'
 #' @export
-annotateDGEobj <- function(dgeObj, regfile, keys = NULL) {
+annotateDGEobj <- function(dgeObj, annotations, keys = NULL) {
 
-    assert_that(file.exists(regfile))
+    if (class(annotations) == "character") {
+        assertthat::assert_that(file.exists(annotations),
+                                msg = "You must provide an annotation text file (annotations) which contains key/value pairs separated by an equals sign.")
```
Comment (cb4ds) [inline R/annotate.R]: Would be good to mention in this message that the file could not be found.  I think we could change this to "The file specified does not exist." or something similar.  What happens if the class of the annotation is not a char?  Ie if someone sends in a list or NULL?

File: R/annotate.R [hunk 2]
```diff
@@ -1,67 +1,78 @@
...
-    regdat$key <- ""
-    regdat$value <- ""
-    for (i in 1:nrow(regdat)) {
-        splitpos <- str_locate(regdat[i,1], "=")[1]  # Pos of 1st = sign
-        regdat$key[i] <- substr(regdat[i,1], 1, (splitpos - 1))
-        regdat$value[i] <- substr(regdat[i,1], (splitpos + 1), nchar(regdat[i,1]))
-    }
+        # Just lines with equals signs
+        regdat <- regdat[grepl("=", regdat$pair) | !grepl("Parameters.", regdat$pair), , drop = FALSE]
+
```
Comment (cb4ds) [inline R/annotate.R:48]: We should also ignore lines that are commented out (ie grepl("^#")) regardless of equals sign

File: R/attributes.R [hunk 3]
```diff
@@ -47,13 +47,13 @@ showAttributes <- function(dgeObj,
...
-#' a names attribute on the attribs argument list that already exists in the
-#' object will be updated.  The function is generic in that it should work on
-#' other data types/classes, not just a DGEobj.
+#' annotation(s) to a DGEobj or to a specific item within a DGEobj.
+#' The setAttributes() function adds the attributes passed to it in
+#' the attribs argument without deleting attributes already present. To remove an attribute,
+#' you can pass NULL as the value. Any named attribute on the attribs argument list
+#' that already exists in the object will be updated.
+#' The function is generic in that it should work on other data types/classes,
+#' not just a DGEobj.
```
Comment (cb4ds) [inline R/attributes.R]: I would remove this mention as we don't want to over-promise without extensive testing.

File: R/conversions.R [hunk 4]
```diff
@@ -73,10 +73,14 @@ convertDGEobj <- function(dgeObj, Class) {
...
-                class(Class) == "character",
-                toupper(Class) %in% supportedClasses)
+
+    assertthat::assert_that(!missing(dgeObj),
+                            !missing(Class),
+                            msg = "Specify both a DGEobj and a desired Class to convert the DGEobj to.")
+    assertthat::assert_that(class(Class) == "character",
+                            msg = "Specify the Class via character string" )
+    assertthat::assert_that(toupper(Class) %in% supportedClasses,
+                            msg = 'The specified class must be one of "RangedSummarizedExperiment", "RSE", "ExpressionSet", or "ES".')
```
Comment (cb4ds) [inline R/conversions.R:83]: Let's incorporate the supportedClasses variable here, this way the message will always match.  ie paste/collapse the values

File: R/conversions.R [hunk 5]
```diff
@@ -150,9 +154,11 @@ convertRSE <- function(RSE,
         return(dgeObj)
     }
 
-    assert_that(!missing(RSE),
-                !missing(Class),
-                class(RSE)[[1]] == "RangedSummarizedExperiment")
+    assertthat::assert_that(!missing(RSE),
+                            !missing(Class),
```
Comment (cb4ds) [inline R/conversions.R:158]: Should we check that the class we are converting to is valid here too?

File: R/get.R [hunk 6]
```diff
@@ -20,11 +20,13 @@
...
-                "DGEobj" %in% class(dgeObj),
-                any(c("character", "list") %in% class(itemNames))
-    )
+    assertthat::assert_that(!missing(dgeObj),
+                            !missing(itemNames),
+                            msg = "Specify both a DGEobj and at least one itemName to retrieve.")
+    assertthat::assert_that("DGEobj" %in% class(dgeObj),
+                            msg = "The DGEobj must be of class 'DGEobj'.")
+    assertthat::assert_that(any(c("character", "list") %in% class(itemNames)),
+                            msg = "Pass the itemNames as a single character string or a list of items to retrieve.")
```
Comment (cb4ds) [inline R/get.R]: I'd reword this slightly to "...or a list of string names to retrieve".  Does it need to be a list or is a character vector also accepted?  If so we need to update the check and message.

File: R/get.R [hunk 7]
```diff
@@ -66,11 +68,15 @@ getItems <- function(dgeObj, itemNames){
...
-                            class(itemName) == "character",
+                            msg = "Specify both a DGEobj and an itemName to retrieve.")
+    assertthat::assert_that("DGEobj" %in% class(dgeObj),
+                            msg = "The DGEobj must be of class 'DGEobj'.")
+    assertthat::assert_that(class(itemName) == "character",
                             length(itemName) == 1,
-                            itemName %in% names(dgeObj)
-    )
+                            msg = "The itemName should be a character string and contain the name of only one item to retrieve.
+                             To retrieve multiple items, use the getItems() function.")
```
Comment (cb4ds) [inline R/get.R]: move to the same line, if you want a line break use \n (I think).

File: R/init.R [hunk 8]
```diff
@@ -88,11 +92,11 @@ initDGEobj <- function(counts,
             sampleCount <- nrow(colData)
             minchar <- nchar(as.character(sampleCount))
             maxchar <- max(sapply(rownames(colData), nchar))
-            assert_that(maxchar > minchar,
-                        msg = str_c("It looks like you have numeric sample IDs (design rownames).",
-                                    "Please supply a more specific sample identifier. ",
-                                    "Use allowShortSampleIDs = TRUE to explicitily override this restriction",
-                                    sep = "\n"))
+            assertthat::assert_that(maxchar > minchar,
+                                    msg = str_c("It looks like you have numeric sample IDs (design rownames).",
```
Comment (cb4ds) [inline R/init.R]: why str_c vs paste?

File: R/reset.R [hunk 9]
```diff
@@ -30,7 +30,9 @@ resetDGEobj <- function(dgeObj, platformType){
     platform.rnaseq <- c("rna-seq", "rnaseq")
 
     assertthat::assert_that("DGEobj" %in% class(dgeObj),
-                            !is.null(attr(dgeObj, "level")))
+                            "The DGEobj must be of class 'DGEobj'.")
+    assertthat::assert_that(!is.null(attr(dgeObj, "level")),
+                            "The DGEobj must have a 'level' attribute specified.")
 
     if (is.null(attr(dgeObj, "PlatformType")))
         stop("Required attribute \"PlatformType\" is missing!  Must use platformType argument.")
```
Comment (cb4ds) [inline R/reset.R]: we do not want to call stop unless absolutely necessary - let's put this check into the assertthat set.

File: R/subset.R [hunk 10]
```diff
@@ -29,7 +29,8 @@
 #' @export
 subset.DGEobj <- function(x, ..., row, col, drop = FALSE, debug = FALSE){
 
-    assertthat::assert_that(class(x)[[1]] == "DGEobj")
+    assertthat::assert_that(class(x)[[1]] == "DGEobj",
+                            msg = "The DGEobj must be of class 'DGEobj'.")
```
Comment (cb4ds) [inline R/subset.R]: update to x - the variable name is not DGEobj for compliance reasons now.

**Review by cb4ds — CHANGES_REQUESTED** [review]

File: R/conversions.R (hunk 4 above)
Comment (klmedeiros-ag) [reply R/conversions.R:83]: The `supportedClasses` variable is in all caps (so "RANGEDSUMMARIZEDEXPERIMENT", etc.) and converting back to correct cases will be challenging, as some are title and some are all caps. I'm not sure of a good way to bring it back in.

**Review by cb4ds — APPROVED** [review]

File: R/conversions.R (hunk 4 above)
Comment (cb4ds) [reply R/conversions.R:83]: ok, no problem, we will likely drop the conversions also.
